cirq-core/cirq/protocols/json_serialization.py

```diff
@@ -338,137 +368,6 @@ class SerializableByKey(SupportsJSON):
     """
 
 
-class _SerializedKey(SupportsJSON):
-    """Internal object for holding a SerializableByKey key.
-
-    This is a private type used in contextual serialization. Its deserialization
-    is context-dependent, and is not expected to match the original; in other
-    words, `cls._from_json_dict_(obj._json_dict_())` does not return
-    the original `obj` for this type.
-    """
-
-    def __init__(self, key: str):
-        self.key = key
-
-    def _json_dict_(self):
-        return obj_to_dict_helper(self, ['key'])
-
-    @classmethod
-    def _from_json_dict_(cls, **kwargs):
-        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
-
-    @classmethod
-    def read_from_context(cls, context_map, key, **kwargs):
-        return context_map[key]
-
-
-class _SerializedContext(SupportsJSON):
-    """Internal object for a single SerializableByKey key-to-object mapping.
-
-    This is a private type used in contextual serialization. Its deserialization
-    is context-dependent, and is not expected to match the original; in other
-    words, `cls._from_json_dict_(obj._json_dict_())` does not return
-    the original `obj` for this type.
-    """
-
-    def __init__(self, obj: SerializableByKey, uid: int):
-        self.key = uid
-        self.obj = obj
-
-    def _json_dict_(self):
-        return obj_to_dict_helper(self, ['key', 'obj'])
-
-    @classmethod
-    def _from_json_dict_(cls, **kwargs):
-        raise TypeError(f'Internal error: {cls} should never deserialize with _from_json_dict_.')
-
-    @classmethod
-    def update_context(cls, context_map, key, obj, **kwargs):
-        context_map.update({key: obj})
-
-
-class _ContextualSerialization(SupportsJSON):
```

Comment:
[No change needed] From our discussion offline: I think the issue with nested objects I was thinking of was linked to this "context table" implementation I used, which did not allow nested definitions. The new design, which allows a `VAL` inside of a `VAL`, should bypass the issue.